File: src/lexer.py
import string
from src.tokens import Token, TokenType

WHITESPACE = ' \t\r\n'
DIGITS = '0123456789'
LETTERS = '_' + string.ascii_letters
KEYWORDS = [
    'if',
    'else',
    'while',
    'function',
    'return',
    'class',
    'enum',
    'extends'
]

ESCAPE_CHARACTERS = {
    'n': '\n',
    't': '\t',
    'r': '\r'
}

class Lexer:
    def __init__(self, path, text):
        self.path = path
        self.text = iter(text)
        self.advance()
    
    def raise_error(self):
        raise SystemExit(f'{self.path}: lexical error')
    
    def advance(self):
        try:
            self.current_char = next(self.text)
        except StopIteration:
            self.current_char = None

    def get_tokens(self):
        tokens = []
        
        while self.current_char != None:
            if self.current_char in WHITESPACE:
                self.advance()
            elif self.current_char in (DIGITS + '.'):
                tokens.append(self.get_number())
            elif self.current_char == "'":
                tokens.append(self.get_char())
            elif self.current_char == '"':
                tokens.append(self.get_string())
            elif self.current_char in (LETTERS + DIGITS):
                tokens.append(self.get_name())
            elif self.current_char == '+':
                self.advance()
                tokens.append(Token(TokenType.PLUS))
            elif self.current_char == '-':
                self.advance()
                tokens.append(Token(TokenType.MINUS))
            elif self.current_char == '*':
                self.advance()
                tokens.append(Token(TokenType.MULTIPLY))
            elif self.current_char == '/':
                self.advance()
                tokens.append(Token(TokenType.DIVIDE))
            elif self.current_char == '<':
                self.advance()
                if self.current_char == '=':
                    self.advance()
                    tokens.append(Token(TokenType.LE))
                else:
                    tokens.append(Token(TokenType.LT)) 
            elif self.current_char == '>':
                self.advance()
                if self.current_char == '=':
                    self.advance()
                    tokens.append(Token(TokenType.GE))
                else:
                    tokens.append(Token(TokenType.GT)) 
            elif self.current_char == '=':
                self.advance()
                if self.current_char == '=':
                    self.advance()
                    tokens.append(Token(TokenType.EE))
                else:
                    tokens.append(Token(TokenType.EQ))
            elif self.current_char == '!':
                self.advance()
                if self.current_char == '=':
                    self.advance()
                    tokens.append(Token(TokenType.NE))
                else:
                    tokens.append(Token(TokenType.NOT))
            elif self.current_char == '&':
                self.advance()
                if self.current_char == '&':
                    self.advance()
                    tokens.append(Token(TokenType.AND))
                else:
                    self.raise_error()
            elif self.current_char == '|':
                self.advance()
                if self.current_char == '|':
                    self.advance()
                    tokens.append(Token(TokenType.OR))
                else:
                    self.raise_error()
            elif self.current_char == ',':
                self.advance()
                tokens.append(Token(TokenType.COMMA))
            elif self.current_char == '(':
                self.advance()
                tokens.append(Token(TokenType.LPAREN))
            elif self.current_char == ')':
                self.advance()
                tokens.append(Token(TokenType.RPAREN))
            elif self.current_char == '{':
                self.advance()
                tokens.append(Token(TokenType.LCURLY))
            elif self.current_char == '}':
                self.advance()
                tokens.append(Token(TokenType.RCURLY))
            else:
                self.raise_error()

        tokens.append(Token(TokenType.EOF))
        
        return tokens
    
    def get_number(self):
        value = self.current_char
        decimal_point_count = 0

        self.advance()

        while self.current_char != None and self.current_char in (DIGITS + '.'):
            if self.current_char == '.':
                decimal_point_count += 1
                if decimal_point_count > 1:
                    break
            
            value += self.current_char
            self.advance()

        if value == '.': 
            return Token(TokenType.DOT)
    
        return Token(TokenType.NUMBER, value)
    
    def get_char(self):
        value = ''
        escape_character = False
        
        self.advance()

        while self.current_char != "'":
            if self.current_char == None:
                self.raise_error()

            if escape_character:
                value += ESCAPE_CHARACTERS.get(self.current_char, self.current_char)
                escape_character = False
            else:
                if self.current_char == '\\':
                    escape_character = True
                else:
                    value += self.current_char
                
            self.advance()
        
        self.advance()

        if len(value) != 1:
            self.raise_error()

        return Token(TokenType.CHAR, value)

    def get_string(self):
        value = ''
        escape_character = False
        
        self.advance()

        while self.current_char != '"':
            if self.current_char == None:
                self.raise_error()
                escape_character = False

            if escape_character:
                value += ESCAPE_CHARACTERS.get(self.current_char, self.current_char)
            else:
                if self.current_char == '\\':
                    escape_character = True
                else:
                    value += self.current_char
                

            self.advance()
            
        self.advance()
        return Token(TokenType.STRING, value)

    def get_name(self):
        value = self.current_char

        self.advance()

        while self.current_char != None and self.current_char in (LETTERS + DIGITS):            
            value += self.current_char
            self.advance()


        if value in KEYWORDS:
            return Token(TokenType.KEYWORD, value)     
               
        return Token(TokenType.NAME, value)

File: lexer.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cstddef>

#include "tokens.h"
#include "lexer.h"

namespace Lexer {
    Lexer::Lexer(std::string filename, std::string text) {
        this->filename = filename;
        this->text = text;
        index = 0;
        advance();
    }

    void Lexer::advance() {
        current_char = text[index++];
        if(current_char == "\n")
            line++;
    }

    std::vector<Tokens::Token> Lexer::get_tokens() {
        std::vector<Tokens::Token> tokens;

        while (index <= text.length()) {
            if (current_char.find_first_of(WHITESPACE) != std::string::npos) {
                advance();
            } else if (current_char == "." || current_char.find_first_of(DIGITS) != std::string::npos) {
                tokens.push_back(get_number());
            } else if (current_char.find_first_of(NONDIGITS) != std::string::npos) {
                tokens.push_back(get_symbol());
            } else if (current_char == "+") {
                advance();
                tokens.push_back(Tokens::Token(line, Tokens::PLUS));
            } else if (current_char == "-") {
                advance();
                tokens.push_back(Tokens::Token(line, Tokens::MINUS));
            } else if (current_char == "*") {
                advance();
                if (current_char == "*") {
                    advance();
                    tokens.push_back(Tokens::Token(line, Tokens::POWER));
                } else
                    tokens.push_back(Tokens::Token(line, Tokens::MULTIPLY));
            } else if (current_char == "/") {
                advance();
                tokens.push_back(Tokens::Token(line, Tokens::DIVIDE));
            } else if (current_char == "(") {
                advance();
                tokens.push_back(Tokens::Token(line, Tokens::LPAREN));
            } else if (current_char == ")") {
                advance();
                tokens.push_back(Tokens::Token(line, Tokens::RPAREN));
            } else if (current_char == "{") {
                advance();
                tokens.push_back(Tokens::Token(line, Tokens::LCURLY));
            } else if (current_char == "}") {
                advance();
                tokens.push_back(Tokens::Token(line, Tokens::RCURLY));
            } else if (current_char == "=") {
                advance();
                tokens.push_back(Tokens::Token(line, Tokens::EQ));
            } else if (current_char == ";") {
                advance();
                tokens.push_back(Tokens::Token(line, Tokens::SEMICOLON));
            } else {
                throw filename + ":" + std::to_string(line) + ": illegal character '" + current_char + "'";
            }
        }

        tokens.push_back(Tokens::Token(line, Tokens::EOF_));

        return tokens;
    }

    Tokens::Token Lexer::get_number() {
        int decimal_point_count = 0;
        std::string number_str = current_char;

        advance();

        while (index <= text.length() && (current_char == "." || current_char.find_first_of(DIGITS) != std::string::npos)) {
            if (current_char == ".")
                if (++decimal_point_count > 1)
                    break;
            
            number_str += current_char;
            advance();
        }

        if (number_str[0] == '.')
            number_str = "0" + number_str;

        if (number_str[number_str.length() - 1] == '.')
            number_str += "0";

        if(decimal_point_count <= 1 && (current_char == "y" || current_char == "Y")) {
            advance();
            return Tokens::Token(line, Tokens::BYTE, (std::int8_t)std::stoi(number_str));
        }

        return Tokens::Token(line, Tokens::NUMBER, std::stod(number_str));
    }

    Tokens::Token Lexer::get_symbol() {
        std::string symbol_str = current_char;

        advance();

        while (index <= text.length() && current_char.find_first_of(DIGITS + NONDIGITS) != std::string::npos) {
            symbol_str += current_char;
            advance();
        }
        
        if(symbol_str == "true") {
            return Tokens::Token(line, Tokens::TRUE);
        } else if(symbol_str == "false") {
            return Tokens::Token(line, Tokens::FALSE);
        } else if(symbol_str == "nil") {
            return Tokens::Token(line, Tokens::NIL);
        } else if(symbol_str == "var") {
            return Tokens::Token(line, Tokens::VAR);
        } else {
            return Tokens::Token(line, Tokens::SYMBOL, symbol_str);
        }
    }

    void print_tokens(std::vector<Tokens::Token>& tokens) {
        for (auto& token : tokens)
            std::cout << token.to_string() << '\n';
    }
}